### PR TITLE
Add extension schema for tooling

### LINKED ISSUE #545: Tech extension: extension definition json schema
we should have a json schema to validate the extension definition

### PR DESCRIPTION
**Pull request by nicolaferraro**
This fixes #545.

I've added the schema as part of the model jar. It can be used as blueprint to define an extension (hence it's called `syndesis-extension-definition-schema.json`) with the tooling. It does not include fields that are not needed when writing the extension (e.g. status, uses and id).

I've included as test the script to generate the base skeleton (prefer to edit it manually for now).

@apupier, @lhein can you have a look?

### CODE REVIEW

**Comment by apupier** [comment]
for #545 , there was also mention of modifying the name of the model file embedded of the Tech Extension jar. It is to allow to create the mapping between the json file and this provided schema easily.
Do you plan to do it in a separate PR?

File: app/rest/model/src/test/java/io/syndesis/model/extension/ExtensionSchemaValidationTest.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (C) 2016 Red Hat, Inc.
```
Comment (apupier) [inline app/rest/model/src/test/java/io/syndesis/model/extension/ExtensionSchemaValidationTest.java]: 2017?

File: app/rest/model/src/main/resources/syndesis-extension-definition-schema.json [hunk 2]
```diff
@@ -0,0 +1,228 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
```
Comment (apupier) [inline app/rest/model/src/main/resources/syndesis-extension-definition-schema.json]: official json schemas examples (for instance http://json-schema.org/example/geo.json ) contains http://json-schema.org/draft-06/schema#

seems that there is also a 07 version released the 19-11-2017 http://json-schema.org/specification.html

not sure what it is changing exactly.
I guess it is best to try to use the latest version first and then if we encounter issues we might choose a specific one, what do you think?

**Comment by nicolaferraro** [comment]
Yes, the issue was about putting the schema in syndesis and we agreed on the name during the discussion. If you need a change also in the model object name, you can open an issue and we'll discuss about that.

**Comment by apupier** [comment]
There is currently no "description" provided on the fields. it would be nice to add them. This can be done in a separate PR/issue.

**Comment by apupier** [comment]
I quickly tried to add the json schema for Eclipse JSON Editor but the schema validation/completion is not working, there is an exception:
```
Exception:org.eclipse.json.provisonnal.com.eclipsesource.json.ParseException: Expected value at 7:0
	 at org.eclipse.json.provisonnal.com.eclipsesource.json.JsonParser.error(JsonParser.java:376)
	 at org.eclipse.json.provisonnal.com.eclipsesource.json.JsonParser.expected(JsonParser.java:369)
	 at org.eclipse.json.provisonnal.com.eclipsesource.json.JsonParser.readValue(JsonParser.java:99)
	 at org.eclipse.json.provisonnal.com.eclipsesource.json.JsonParser.parse(JsonParser.java:64)
	 at org.eclipse.json.provisonnal.com.eclipsesource.json.JsonValue.readFrom(JsonValue.java:91)
	 at org.eclipse.json.provisonnal.com.eclipsesource.json.JsonObject.readFrom(JsonObject.java:122)
	 at org.eclipse.json.impl.schema.JSONSchemaDocument.<init>(JSONSchemaDocument.java:29)
	 at org.eclipse.wst.json.schemaprocessor.internal.JSONSchemaProcessor.getSchema(JSONSchemaProcessor.java:55)
	 at org.eclipse.wst.json.core.internal.schema.SchemaProcessorRegistryReader.getSchemaDocument(SchemaProcessorRegistryReader.java:106)
	 at org.eclipse.wst.json.core.internal.validation.eclipse.Validator.validate(Validator.java:131)
	 at org.eclipse.wst.json.core.internal.validation.eclipse.Validator.validate(Validator.java:110)
	 at org.eclipse.wst.json.core.internal.validation.core.AbstractNestedValidator.validate(AbstractNestedValidator.java:281)
	 at org.eclipse.wst.json.core.internal.validation.core.AbstractNestedValidator.validateInJob(AbstractNestedValidator.java:152)
	 at org.eclipse.wst.json.ui.internal.validation.DelegatingSourceValidator.validate(DelegatingSourceValidator.java:237)
	 at org.eclipse.wst.sse.ui.internal.reconcile.validator.ReconcileStepForValidator.validate(ReconcileStepForValidator.java:250)
	 at org.eclipse.wst.sse.ui.internal.reconcile.validator.ReconcileStepForValidator.reconcileModel(ReconcileStepForValidator.java:208)
	 at org.eclipse.jface.text.reconciler.AbstractReconcileStep.reconcile(AbstractReconcileStep.java:89)
	 at org.eclipse.wst.sse.ui.internal.reconcile.validator.ValidatorStrategy.reconcile(ValidatorStrategy.java:269)
	 at org.eclipse.wst.sse.ui.internal.reconcile.DocumentRegionProcessor.process(DocumentRegionProcessor.java:321)
	 at org.eclipse.wst.sse.ui.internal.reconcile.StructuredRegionProcessor.process(StructuredRegionProcessor.java:258)
	 at org.eclipse.wst.sse.ui.internal.reconcile.DirtyRegionProcessor$BackgroundThread.run(DirtyRegionProcessor.java:691)
```

I have no idea from what it comes yet (if it is an issue with the schema or with the JSON editor or an incompatibility for any reason between the provided schema and the JSON Editor)

to reproduce:
- use JBoss Tools
- Window -> preferences -> JSON -> JSON Catalog
- Add:
-- File match:  extension-definition.json
-- URL: point to the schema
- try to open an extension-definition.json file with the JSon Editor
-6> the erro ris only in log but there is no completion/validation available

**Comment by lburgazzoli** [comment]
@nicolaferraro  could you update this pr to reflect #714 ?

File: app/rest/model/src/main/resources/syndesis-extension-definition-schema.json (hunk 2 above)
Comment (nicolaferraro) [reply app/rest/model/src/main/resources/syndesis-extension-definition-schema.json]: Most Java parsers are compatible with draft-03 and draft-04, that's why I've used that version. Better checking the eclipse tooling compatibility

File: app/rest/model/src/test/java/io/syndesis/model/extension/ExtensionSchemaValidationTest.java (hunk 1 above)
Comment (nicolaferraro) [reply app/rest/model/src/test/java/io/syndesis/model/extension/ExtensionSchemaValidationTest.java]: The syndesis license header plugin requires 2016. We should update all headers (better to 2018 :D) in another PR.

**Comment by nicolaferraro** [comment]
Yes @lburgazzoli, I'll change the path.

@apupier, before I install JBoss Tools tomorrow (I've a fresh laptop) and check the issue, can you try (e.g. with a reduced version of that schema) to understand what is causing the issue?

**Comment by nicolaferraro** [comment]
@apupier I'm not able to reproduce it following your steps. I'm using Dev Studio 11 with JBoss Tools. I'm checking the console log. Is there any other place where dev studio stores the logs?

**Comment by nicolaferraro** [comment]
@apupier I managed to configure it (validation and suggestion should be enabled in the configuration) and everything works.

img_20171213_152038

It may be a bug of your version of the IDE. I'll just move the file to the syndesis directory for now.

**Comment by lburgazzoli** [comment]
@nicolaferraro @apupier can we merge this ?

**Comment by oscerd** [comment]
+1

**Comment by rhuss** [comment]
Will be merged as soon as the the "wip" label is removed ....

**Comment by nicolaferraro** [comment]
I think we can safely merge this so we can update quickstarts to use the new descriptor name. We can address other issues later.

**Comment by apupier** [comment]
> It may be a bug of your version of the IDE. I'll just move the file to the syndesis directory for now.

it was a bug of the user (me), I used the wrong URL to the schema on github (not the "raw" one)

so it is working, that's great!

**Pull request merged**